core/src/main/scala/kafka/server/ControllerApis.scala

```diff
@@ -117,6 +117,11 @@ class ControllerApis(val requestChannel: RequestChannel,
           s"with context ${request.context}", t)
         requestHelper.handleError(request, t)
       }
+    } finally {
+      // Only record local completion time if it is unset.
+      if (request.apiLocalCompleteTimeNanos < 0) {
+        request.apiLocalCompleteTimeNanos = time.nanoseconds
```

Comment:
All the requests to the controller and kraft get queued for another thread to do the work. That means that for these RPCs the api local time is the time it took to authorize the request and queue the request for the controller or kraft, right?

Reply:
Right. The asynchronous processing time is counted as "remote time." Most of the time, we are waiting for replication of some kind, so the meaning is mostly correct. The "local time" is mainly measuring how much time was spent on the request threads. Perhaps we could consider an additional level of granularity which accounts for the time spent on the raft/controller threads. Seems like that would be useful.